src/librustc_save_analysis/dump_visitor.rs

```diff
@@ -1233,10 +1233,13 @@ impl<'l, 'tcx: 'l, 'll, O: DumpOutput + 'll> DumpVisitor<'l, 'tcx, 'll, O> {
     fn process_use_tree(&mut self,
                          use_tree: &'l ast::UseTree,
                          id: NodeId,
-                         parent_item: &'l ast::Item,
+                         root_item: &'l ast::Item,
                          prefix: &ast::Path) {
         let path = &use_tree.prefix;
-        let access = access_from!(self.save_ctxt, parent_item);
+        let access = access_from!(self.save_ctxt, root_item);
+        let parent = self.save_ctxt.tcx.hir.opt_local_def_id(id)
+            .and_then(|id| self.save_ctxt.tcx.parent_def_id(id))
+            .map(::id_from_def_id);
```

Comment:
I'm not clear on what the tcx regards as the parent for a nested import - is it the enclosing module? Or the outer import? It seems like Rustdoc always wants the former. Which I would imagine is either the root_item or the parent of the root_item. But perhaps `id` is the `id` of the root item? Or the tcx always gives the module?

Anyway, could you explain what you think is going on here please?

Reply:
cc @pietroalbini, who might have some ideas on how to best handle this

Reply:
The `parent_def_id` of a nested import should be the parent of the enclosing thing, regardless of the nesting level ([no defs are added](https://github.com/rust-lang/rust/blob/3fc7f8522bbe59189b6423295a65f337b9735a88/src/librustc/hir/map/def_collector.rs#L121-L123) for the `use` item, and [no extra parents are set](https://github.com/rust-lang/rust/blob/3fc7f8522bbe59189b6423295a65f337b9735a88/src/librustc/hir/map/def_collector.rs#L170-L173) while creating the use tree's defs).

I don't see why the parent imports should be stored, since each import already has the full path in it, but I'm not familiar with rls at all so I don't know how it uses those things.

Reply:
I think that logically rustdoc wants a list imports and the module they are imported into. I.e., the imports are flattened for the RLS.

The `parent` here is the module (or function or block, whatever) where the import is declared and is the context into which the imported name is added. It's not 'parent import'.

So, it sounds like we are doing the right thing here.